### DISCUSSION TITLE
Inav (reference 3.0 or earlier)  Mission warm in flight reset/restart from scratch

### DISCUSSION THREAD

**Discussion by alccrl** [iNavFlight/inav#7051] [Q&A] [answered]
Hi,

I tried to restart from scratch (the beginning WP1) a mission that was suspended in flight, but i got always LANDING. This could be related to last WP that was of landing type, and very next to the first WP, so, I guess, mission restarts frome the nearest waypoint (actually the takeoff point that is equal to the return last point).
But I am not sure this one explanantion ....

Is there any way to restart a mission, with no need to reload a different file to the controller/eeprom ?
i did not find any doc or video referring this situation.

Alfredo

**Comment by breadoven** [comment, marked as answer]
WP missions always start from WP 1 regardless of being disrupted mid mission. The OSD system message should show which WP number is active when you enable WP mode, should be WP 1. Something odd going on if it isn't. Are you sure this is mission landing or landing for some other reason ?

**Reply by alccrl** [reply]
Mission is a polygon with last point next to first point (takeoff/landing point).
If I interrupt mission (back to POS_HOLD) , and then I try to start it again using stick commands, I get the drone trying to land.
I get these messages (attached), I realized now, these may help.
![Screenshot from VID00002 AVI](https://user-images.githubusercontent.com/78967144/120366156-9cf28f80-c30f-11eb-98a3-91919d65b269.png)
![Screenshot from VID00002 AVI - 1](https://user-images.githubusercontent.com/78967144/120366161-9e23bc80-c30f-11eb-9f12-34012f397b48.png)

**Comment by alccrl** [comment]
Before this test I made another one - got same behaviour, but in goggles view there was an osd shift down of one line - may be another bug or something like that either in osd or in goggles, so I lost system messages (see screenshot).
![Screenshot from VID00002 AVI - 2](https://user-images.githubusercontent.com/78967144/120371502-3755d180-c316-11eb-83d1-ba8cfd72f389.png)

**Reply by breadoven** [reply]
OK. The messages show it's in RTH mode and since you're already at home it lands immediately, hence the Landing. If you select WP mode and there is no mission loaded or it's invalid navigation falls back to RTH mode. So it would appear there is some problem with the mission not loading properly. Not sure why you mention stick commands when restarting the mission. Stick commands only load/unload/save a mission they can't start it. To restart the mission just use the WP mode switch and it will restart the previously loaded mission.

**Reply by alccrl** [reply]
When I refer "stick commands" I refer a rc switch (knob) configured for Wp mode on/off. I have a switch (knob) for RTH too, but I am sure their range do not interfere, they are separate knobs in Flysky transmitter. I use also mode 2 stick commands for loading mission (this is whay You refer).  By the way .... loading mission I fall back (I guess) in issue #6987. ..
I have also blackbox logs, if they can help, and complete cli setup dump. RTH is my failsafe choice, but drone does not show any problem, at the moment, and it was very next to the RC while trying to restart mission. In outdooring sometime would be more feasable to "join" the route (the mission), not to restart from scratch from first (may be far) wp. if this happens to Inav, it means that it joins last point and it lands, but it seems it not should work like this, as You explains.

**Reply by breadoven** [reply]
Is the last WP a LAND WP or a RTH with landing WP ?
Did the Arming screen show a message "MISSION LOADED" ?
Is nav_wp_load_on_boot ON and if so are you using the Mission load stick command before restarting the mission ? If nav_wp_load_on_boot is ON and you use the stick command to load a mission when a mission is already loaded than the stick command will actually unload the mission (toggles between loading and unloading). This will cause WP mode to fall back to RTH because no mission is loaded.

The CLI "WP" output would be useful as would a DIFF file.

Also there is https://github.com/iNavFlight/inav/pull/6938 which will allow missions to be resumed from the last active WP if the mission is interrupted. Not merged yet though.

**Reply by alccrl** [reply]
I think I fell in the wrong "toggle" situation, I did not know that the mission command was actually a toggle !
I have been searching docs about these details but I did not find anything.
Goggles recording shows MISSION LOADED only once (first mission start) so You are right, while playing with mission stick command may be I unloaded it :-).
I attach the mission dump from eeprom, I am sorry, I am not 100 % sure that this is the one I used, because it happened a few days ago, may be I by mistake overwrote it, but since I used android mission planner that does not distinguish waypoints types, because it is not updated, and I checked the mission against inav configurator and mwp last versions, much probably I did not use LANDING or RTH waypoints, in order not to repeat my first test (landing instead of restart) result.
Summarizing. it seems that I tried to start an unloaded mission :-), because I did not know that stick command is a toggle.

Other then source code, is there any document that explains mission cases ?

Alfredo
[INAV_cli_20210602_113536.txt](https://github.com/iNavFlight/inav/files/6583557/INAV_cli_20210602_113536.txt)

**Reply by breadoven** [reply]
OK, so it seems no mission was loaded which explains the RTH. And actually I was wrong that `nav_wp_load_on_boot ` needs to be enabled to enable the WP load/unload toggle. The toggle is now set by default. Perhaps it should only toggle if `nav_wp_load_on_boot `is ON ... or it needs documenting more clearly. See also https://github.com/iNavFlight/inav/issues/7005. @tonyyng 

The following explain missions:
https://github.com/iNavFlight/inav/blob/master/docs/Navigation.md#cli-command-wp-to-manage-waypoints
https://github.com/iNavFlight/inav/wiki/MSP-Navigation-Messages

**Reply by tonyyng** [reply]
> And actually I was wrong that `nav_wp_load_on_boot ` needs to be enabled to enable the WP load/unload toggle. The toggle is now set by default. Perhaps it should only toggle if `nav_wp_load_on_boot `is ON ... or it needs documenting more clearly.

I don't like the idea of toggling only when the `nav_wp_load_on_boot = ON`.  Maybe there should be a message when a mission is unloaded.  I'll look into that.  Is there a spot in the documentation that you think should be made more clear about the toggling behavior?

**Reply by breadoven** [reply]
> > And actually I was wrong that `nav_wp_load_on_boot ` needs to be enabled to enable the WP load/unload toggle. The toggle is now set by default. Perhaps it should only toggle if `nav_wp_load_on_boot `is ON ... or it needs documenting more clearly.
> 
> I don't like the idea of toggling only when the `nav_wp_load_on_boot = ON`. Maybe there should be a message when a mission is unloaded. I'll look into that. Is there a spot in the documentation that you think should be made more clear about the toggling behavior?

You already made a change in https://github.com/iNavFlight/inav/blob/master/docs/Controls.md although it didn't include changing the diagram of the stick commands which maybe it should given that's probably what people look at as a reference more than the table. Perhaps add a footnote clarifying the toggle behaviour and different beeper sound between loaded and unloaded.

Also https://github.com/iNavFlight/inav/blob/master/docs/Navigation.md#cli-command-wp-to-manage-waypoints mentions WP loading from the CLI. I noticed the toggle unload also works in the CLI which could be confusing if you're not aware.

**Reply by alccrl** [reply]
OK, Thanks !
As soon as possible I will read "MSP Navigation Messages" in the Wiki and retest.

If You change the code or docs, I guess I will see the list of changes somewhere in github ...
(I already had a look at the source code and compiled, I was curious, it (2.6.1 for mateksys-f7-std), but I have yet to find a good way to navigate it. Eclipse CDT seems to be supporting docker systems, but since I do not use docker systems, I just configured in Eclipse some target compilation #defines looking at generated cmakefiles, in order to correctly navigate active code).
  
Alfredo

**Reply by tonyyng** [reply]
> You already made a change in https://github.com/iNavFlight/inav/blob/master/docs/Controls.md although it didn't include changing the diagram of the stick commands which maybe it should given that's probably what people look at as a reference more than the table. 

There are three files that make up the diagram. 
- StickPositions.ong
- StickPositions.svg
- StickPositions_trans.png

@shellixyz I don't want to make a mess of your changes in these files.  Can you advise me how to make these changes and what tools I need to use?  The minimum change would be to replace "Load waypoint mission" with "Load/unload wp mission".  I could also change "Save waypoint mission" with "Save wp mission" to make the two labels consistent.  I wouldn't object if you were to just give me the updated files. :-)

> Perhaps add a footnote clarifying the toggle behaviour and different beeper sound between loaded and unloaded.
The sound should already be different:

```// Load waypoint list
if (rcSticks == THR_LO + YAW_CE + PIT_HI + ROL_HI) {
    const bool success = loadNonVolatileWaypointList();            <- returns false when unloading
    beeper(success ? BEEPER_ACTION_SUCCESS : BEEPER_ACTION_FAIL);
}
```

**Reply by tonyyng** [reply]
> Also https://github.com/iNavFlight/inav/blob/master/docs/Navigation.md#cli-command-wp-to-manage-waypoints mentions WP loading from the CLI. I noticed the toggle unload also works in the CLI which could be confusing if you're not aware.

Fixed

**Reply by breadoven** [reply]
> > Perhaps add a footnote clarifying the toggle behaviour and different beeper sound between loaded and unloaded.
> > The sound should already be different:
> 
> ```
> if (rcSticks == THR_LO + YAW_CE + PIT_HI + ROL_HI) {
>     const bool success = loadNonVolatileWaypointList();            <- returns false when unloading
>     beeper(success ? BEEPER_ACTION_SUCCESS : BEEPER_ACTION_FAIL);
> }
> ```

The sounds are already different, However, before you only got BEEPER_ACTION_FAIL if no mission was stored in EEPROM, otherwise it just kept repeating BEEPER_ACTION_SUCCESS. Now it toggles between the 2 depending on whether the mission has loaded or unloaded which tells you what the stick command just did.
